README
This is version 2.11 of ``noweb'', a simple, extensible literate-
programming tool.  noweb is available by anonymous ftp from
ftp.eecs.harvard.edu:pub/nr/noweb and by HTTP from
http://www.eecs.harvard.edu/nr/noweb/dist.  The ftp site is mirrored
by the Comprehensive TeX Archive Network, in directory web/noweb.
The home page at http://www.eecs.harvard.edu/~nr/noweb contains much
helpful information.

To build noweb, change to the `src' directory and follow the
instructions in the file INSTALL.


WHAT'S NEW IN VERSION 2.11

Version 2.11 fixes a surprising bug in noweb's parser and provides a
dozen or so small improvements.


INTRODUCTION --- WHAT IS NOWEB, ANYWAY?

noweb is designed to meet the needs of literate programmers while
remaining as simple as possible.  Its primary advantages are
simplicity, extensibility, and language-independence.  noweb uses 5
control sequences to WEB's 27.  The noweb manual is only 3 pages; an
additional page explains how to customize its LaTeX output.  noweb
works ``out of the box'' with any programming language, and supports
TeX, latex, and HTML back ends.  A back end to support full hypertext
or indexing takes about 250 lines; a simpler one can be written in 40
lines of awk.  Noweb does not prettyprint natively, but supports
prettyprinting through such third-party filters as pretzel and dpp.

noweb has been distributed free of charge for ten years, and it is
one of the world's most widely used literate-programming tools.  It
has been used for hundreds of thousands of lines of code in such
languages as awk, C, C++, Icon, Modula-3, PAL, perl, Promela, and
Standard ML.

If you already know you want to use noweb, you need only install it
and read the manual page.  If you're just curious about noweb, read
the paper that appeared in the September 1994 issue of IEEE Software.
(If you can't get Software, send me a postcard and I'll send you a
reprint.)  A nice, brief tutorial appeared in the October 1997 issue
of Linux Journal (but beware that chunk syntax is <<name>> and not <<name>).

If you're brand new to literate programming, check out the FAQ for the
USENET newsgroup comp.programming.literate.   There are also some
resources available through the noweb home page:
   http://www.eecs.harvard.edu/~nr/noweb


WHAT YOU GET IN THIS DISTRIBUTION

This distribution contains the following directories:
  binaries	Pre-built distributions containing binaries and documentation
  contrib	software contributed by noweb users 
  examples	sample noweb filters and programs in different languages
  src		Source code and documentation (including FAQ) for noweb
Where appropriate, these directories have README files of their own.
BINARY DISTRIBUTIONS MAY NOT ALWAYS BE UP-TO-DATE, especially DOS
distributions.
  

EXTENSIBILITY

noweb provides extensibility by using the Unix toolkit philosophy.
The ``noweb,'' ``notangle,'' and ``noweave'' commands are built from
pieces, which are then assembled in pipelines using shell scripts.
The pieces include:

  markup	convert noweb file from human syntax to tool syntax
  unmarkup	inverse of markup
  totex		convert from tool syntax to TeX/latex markup
  tohtml	convert from tool syntax to HTML markup
  nt		`tangle' the tool form of the noweb file
  mnt		discover roots, then act like nt
  noidx		insert indexing and cross-reference information
  finduses	finds uses of identifiers

These pieces are combined by the scripts in the src/shell directory to
provide more than just weaving and tangling:

  noweb		analog of nuweb
  notangle	analog of TANGLE
  noweave	analog of WEAVE
  nountangle	tangle, but keep interleaved documentation in comments
  noroots	print names of all root chunks in a noweb file
  nocount	count number of lines of code and documentation.
  nodefs	extract defined identifiers for noweave -indexfrom
  noindex	build an external index for multi-file documents


WEAVING --- A TAR PIT

The worst aspect of literate programming is the enormous amount of
time wasted wrangling over what ``woven'' output should look like.
Although noweb does no prettyprinting, it is not entirely immune---
several people have complained about noweave's output or have sent me
changes that add more options to noweave.  I resisted for years, but
with version 2.5 I finally succumbed.  I let the number of options to
noweave double, and I have provided for too many options and hooks for
customizing the latex output.  I won't let it happen again.

noweb doesn't try to be all things to all programmers, but it is very
easy to change.  If you don't like noweave's formatting, you can read
tex/support.nw to learn how to customize it; look for the words
``style hook.''  (Reading noweb.sty directly is not recommended.)  For
simple formatting, it might be easier to throw away noweave and make
your own.  To help you get started, the shell directory contains
noweave.simple, a simplified version of noweave that Dave Hanson
created for use with C programs (it can't handle code with @ signs).
The Noweb Hacker's Guide (xdoc/guide.tex) explains the intermediate
language that noweb uses to represent literate programs.

The intermediate language makes it possible to extend noweave with a
language-dependent prettyprinter, as shown by contributions of an Icon
prettyprinter by Kostas Oikonomou and a guarded-command prettyprinter
by Conrado Martinez-Parra.  (I haven't written a prettyprinter myself
because my experience with Spider left me thinking that prettyprinting
is far more trouble than it's worth.)  Further contributions of
prettyprinters are welcome.

noweb comes with two cross-referencers for use with noweave.  The
standard one is written in Icon, because Icon offers excellent
functionality and performance.  Because Icon is not available on all
platforms, I profide an alternative, but inferior cross-referencer
written in awk.  See the INSTALL file for details.

Cross-referencing makes formatting even more of a tar pit; the
cross-referencer itself takes about 300 lines, and extensive LaTeX
support is also required.  I haven't made the attempt to write
cross-reference code for plain TeX.  Anyone who has ideas for reducing
the number of options or for other ways to restore sanity to the
situation is urged to write to nr@eecs.harvard.edu.


INDEX AND IDENTIFIER CROSS-REFERENCE

Noweb creates identifier cross-references so that you can click on an
identifier and jump to its definition (if you're using printed LaTeX
output, a footnote gives you the page number of the definition).  
To noweb, any string of nonwhite characters can be an identifier.  
A human being or a language-dependent tool must mark definitions of
identifiers; noweb finds the uses using a language-independent
algorithm.  The algorithm relies on an idea taken from the lexical
conventions of Standard ML.  Characters are divided into three
classes: alphanumerics, symbols, and delimiters.  If an identifier
begins with an alphanumeric, it must be delimited on the left by a
symbol or a delimiter.  If it begins with a symbol, it must be
delimited on the left by an alphanumeric or a delimiter.  If it begins
with a delimiter, there are no restrictions on the character
immediately to the left.  Similar rules apply on the right-hand side.
The default classifications are chosen to make sense for commonly used
programming languages, so that noweb will not recognize `zip' when it
sees `zippy', or `++' when it sees `++:='.  This trick works
surprisingly well, but it does not prevent noweb from spotting
identifiers in comments or string literals.

The basic assumption in noweb is that a language-dependent filter will
identify definitions automatically.  Filters for Icon, TeX, and yacc
all take about 30 lines of Icon code and are included in the noweb
distribution; if you write a filter of your own, you can put it in the
$LIB directory with a name like `autodefs.pascal'.  Try `noweave
-showautodefs' for a complete list of such filters. 

If the automatic filter does not work well for you, or if there is no
filter available for your language, I recommend that you mark
definitions using backticks (`) in your source code, and use
`-filter btdefn' with both noweave and notangle.

noweave -index works well for short programs, but nodefs, noindex, and
noweave -indexfrom are there for large multi-file programs.  See the
noindex man page for details.


NOTES

  src/xdoc/techrep.* contains an early, almost unrecognizable version
  of a paper describing noweb that appeared in IEEE Software in
  September, 1994.  You are probably better off writing me for a
  reprint of the Software paper.  (Send a postcard!)

  The Noweb Hacker's Guide, which appears in src/xdoc/guide.*,
  documents the representation of noweb files that is used by the
  noweb tools, in case you want to write any tools of your own.
  Simple tools (e.g. count the number of lines of interleaved
  documentation) are trivial.  If you write any tools, or you want
  tools written (e.g. prettyprinters, index generators), let me know.

  The icon directory contains Icon versions of most pipeline
  stages.  If you want to adapt noweb to work with a text processor
  other than TeX or latex, they might provide a better starting point.
  Perhaps the whole system should have been written in Icon from the
  beginning, but I'm not going to do it over.  Icon is available by
  anonymous ftp from cs.arizona.edu.

  Phil Bewig has adapted noweb to troff, and as soon as we get the
  portability bugs ironed out, I'll include that code in the noweb
  distribution.

  
  Thanks to Preston Briggs for the Aho-Corasick recognizer, and for
		helpful discussions.
  Thanks to Dave Hanson for cpif.
  Thanks to Dave Love for LaTeX wizardry.
  Thanks to Joseph Reynolds for prodding me to fix [[...]].
  Thanks to Bill Trost for the original HTML back end.
  Thanks to Philip Miller and Lee Wittenberg for DOS binaries.
  Thanks to Paolo Ciccone for the Win32 binaries.

Send comments or questions to nr@eecs.harvard.edu.  

I enjoy hearing from noweb users; if you have enjoyed noweb, why not
send me a local postcard for my collection?  My address is:

  Norman Ramsey
  Division of Engineering and Applied Sciences
  Maxwell Dworkin 231
  Harvard University
  33 Oxford Street
  Cambridge, MA 02138
               USA


COPYRIGHT

Noweb is copyright 1989-2006 by Norman Ramsey.  All rights reserved.

You may use and distribute noweb for any purpose, for free.  You may
modify noweb and create derived works, provided you retain the
copyright notice, but the result may not be called noweb without my
written consent.  You may not sell noweb itself, but you may do
anything you like with programs created with noweb.


CHANGES
CHANGES FOR VERSION 2.11 - released 6 April 2006
Braces did not render correctly in [[<<$\{a_i\}$>>]] (Albert Reiner)
From Kevin Quick, Windows hints and a new NOTANGLE.BAT
Generate postscript with -P cmz for better PDF (Francky Leyn)
Clarify for noweave -delay that the first doco chunk is before the first @
extended xchunks with -summary argument
Line in a code chunk opening with unclosed, unescaped << is OK (Ralf Hemmecke)

UNDO ME!  Make docs2comments filter tolerate embedded @xref stuff..
Added Windows XP help to FAQ (David Lemmon)
Noidx now calculates labels for a code chunk using the number of preceding
  definitions from the same @file, not the overall number
Change wording for \noweboptions{german} (Christian Lindig)
Need to insert \leavevmode in \sublabel so that new \@bsphack will work right
From Yann Dirson, new filters in contrib/ydirson.
From Fr'ed'eric Lin'e, some improvements in French translation (apologies to Moli`ere)

CHANGES FOR VERSION 2.10c - released 17 Jul 2002

Bug in noidx.nw caused tag NWD1 etc to be used in multiple files.
Cleaned up alignment tags in tables emitted by l2h.nw (Federico Di Gregorio)
Fixed some bugs in the HTML markup for the hacker's guide
Goes back to iconc, except for noidx.nw, which can't be compiled because
  of the 32-bit polynomial arithmetic

CHANGES FOR VERSION 2.10b - released 15 Jul 2002

Bug in xchunks.nw, around line 100 -- db never initialized (Stephen Edwards)
In local identifier cross-reference, enable breaks between pages 
   of multiply defined identifiers (option breakdefs)
Create a secret \noweboptions{breakcode} for those who want their
   code chunks broken across page boundaries
German is `Benutzt' not `Benutztt' (Johannes Huesing)
No longer suggests iconc; compiling with icont is now the default.

CHANGES FOR VERSION 2.10a - released 30 Mar 2001

Fixed problem in src/Makefile -- FAQ should be built by `source' not `all'

CHANGES FOR VERSION 2.10 - released 16 Jun 2000, but not officially

Added -matlab to nountangle (Uwe Brauer)
Added \nwblindhyperanchor to \weblabel (Andrew Swann)
Changed one more 'char c' to 'unsigned char c' in recognize.nw (Gilles ANDRE)
Change noweb.sty to support more than 26 code chunks on a page (F. Leyn)
l2h skips optional arguments with \includegraphics
fixed bugs in htmltoc having to do with nested anchors on one line
Added capacity for two more filters to noweave script
Fixed bug in autodefs.sml (including spaces in symbol chars)
Added autodefs.mmix
Added autodefs.asdl
Scripts for Plan 9 rc (Russ Cox)
Tinkered with noidx sortkey
Added \mbox{} to \nwnewline, so that spaces won't be eaten (John Cummings)
Added xchunks filter, per request of Mike Smith & Glenn Holloway
Replaced FORTH uniqueid heuristic with 32-bit CRC
Added pipedocs filter to pipe documentation chunks through commands
Added h2a (convert html to ascii) for use with pipedoces
Added docs2comments filter, with pipedocs and h2a, to replace nountangle
Added nwmtime filter; helps derive Zephyr web pages
Moved the official FAQ to html
Added multiple-language support to noweb.sty (Miguel Filgueiras)
notangle -L initial newline is a bug and will never be fixed (Harald Soleng)
Added \noweboptions{noxref}, intended for temporary use inside documents.

CHANGES FOR VERSION 2.9a - released 16 Feb 1999

Fixed bug in htmltoc that left multiple toc anchors


CHANGES FOR VERSION 2.9

Good support for hypertext using the `hyperref' package
noweave now finds uses in quoted code, the better to exploit hypertext
added support for emacs font-lock highlighting as comments to noweb-mode (Adnan Yaqub)
Document common -L options on notangle man page (Brett Presnell)
Fixed incompatibility with verbatim package (Rafael Laboissiere)
added contrib/leyn from Francky Leyn
noweb-mode.el no longer treats @x as chunk end for non-white x (Chris Milner)
Don't use signed char to index array of 256 in recognize.nw (Michael Herrmann)
Added explanation for WNT to FAQ.
Support \ quoting in argument to elide
Support multiple -R options on the same notangle command line.
Change tohtml so that noweave -latex+html doesn't protect anchor texts
   in a rawhtml environment (???)
Added to /usr/doc/noweb for Slackware binary distribution
Slackware binary distribution now includes emacs noweb mode.
Updated internal docs in support.nw to explain the great \nwlbrace trouble
Use gcc -ansi -pedantic in Makefiles.
Remove multicols environment from examples/wc.nw (Yotam Medini)
Change noweave and totex to support subscripting identifiers used in quoted code
Fixed bug in l2h (bad conversion of tables).
Make l2h of \section use CPS (for newline between \section and {)
Added letenv to l2h, and changed verbatim to use environment name as terminator


CHANGES FOR VERSION 2.8a

noweave complains if you use -x with -index or -indexfrom
installation now includes noweb-mode.el
noroots no longer treats uses in quoted code as real uses (David Bruce)
file INSTALL now notes that 'awkname' is always needed (Robert Solovay)
file INSTALL now notes that texhash(1) may be needed (Robert Solovay)
mnt (noweb script) now complains about chunks named <<*>> (Robert Solovay)
lots of rcs identifiers and names everywhere
all man pages should carry version numbers
Use Stephen Parker's htmltoc, with many bug fixes and improvements.


CHANGES FOR VERSION 2.8

fixed unbalanced brace in support.nw for nwmac.tex (Hartmut Henkel)
changed notangle script so notangle -L preserves tabs (Saul London, NR)
changed uniqueid in noidx to avoid everything except letters, digits,
	and a few known special chars.  fixes the Norwegian bug (Hans
	Husum) and also the non-Posix regexp (Christian Wagner)
update recognizer.nw to use 128-node recognizer for larger character sets
	as per Christian Wagner
$arkopts -> $markopts for noweave -v
noweave now uses \documentclass for headers
fixed a bug in tohtml ``impossible character to skip''
hacked \catcode \other for " in \setupcode, 

src/c/markmain.c
#line 66 "markmain.nw"
static char rcsid[] = "$Id: markmain.nw,v 2.24 2006/04/02 16:45:17 nr Exp nr $";
static char rcsname[] = "$Name: v2_11 $";
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <assert.h>
#include "errors.h"
#include "markup.h"
#include "getline.h"
#include "columns.h"

#line 35 "markmain.nw"
typedef enum state {Code=1, Docs=2, CodeIndex=3} State;
typedef enum mark {Begin=1, End=2} Mark;
typedef enum index {Defn=1, Use=2, Newline=3} Index;

static char *states[]  = { "bad state", "code", "docs", "code" };
static char *marks[]   = { "bad mark", "begin", "end" };
static char *indices[] = { "bad index", "defn", "use", "nl" };
static char low_at_sign = '@';

static void print_state(FILE *out, Mark be, State state, int count) {
    fprintf(out, "%c%s %s %d\n", low_at_sign, marks[be], states[state], count);
}

static void print_index(FILE *out, Index idx, char *arg) {
    if (arg)
        fprintf(out, "%cindex %s %s\n", low_at_sign, indices[idx], arg);
    else
        fprintf(out, "%cindex %s\n",    low_at_sign, indices[idx]);
}

static void print_pair(FILE *out, char *name, char *value) {
    if (value) {
        int last=strlen(value)-1;
        if (last>=0 && value[last]=='\n')
            fprintf(out, "%c%s %s%cnl\n", low_at_sign, name, value, low_at_sign);
        else
            fprintf(out, "%c%s %s\n", low_at_sign, name, value);
    } else
        fprintf(out, "%c%s\n", low_at_sign, name);
}

#line 80 "markmain.nw"
void markup (FILE* in, FILE *out, char *filename) {
    State state = Docs;         /* what we are reading */
    int quoting = 0;            /* currently quoting code? */
    int count = 0;              /* number of current chunk, from 0 */
    int missing_newline;        /* was last line missing a trailing \n? */
    int lineno = 0;             /* number of lines read */
    int last_open_quote;        /* line number of last unmatched open quote */

    char *line;                 /* part of line up to mark (or end) */
    char *mark;                 /* potentially marked part of line */
#define MAX_MODNAME 255
    char modname[MAX_MODNAME+1] = ""; /* name of module currently being read, 
                                         [[""]] if no module is being read */ 

    
#line 121 "markmain.nw"
print_pair(out, "file", filename);
print_state(out, Begin, state, count);
while ((line = getline_expand(in)) != NULL) {
    lineno++;
    
#line 172 "markmain.nw"
missing_newline = line[strlen(line)-1] != '\n';
#line 126 "markmain.nw"
    if (starts_code(line, filename, lineno)) {
        
#line 542 "markmain.nw"
if (quoting) {
    errorat(filename, last_open_quote, Warning, "open quote `[[' never closed");
    quoting = 0;
}
#line 128 "markmain.nw"
        print_state(out, End, state, count);
        count++;
        state = Code;
        print_state(out, Begin, state, count);
        getmodname(modname,MAX_MODNAME,line);
        print_pair(out,"defn",modname);
        print_pair(out,"nl",0);     /* could be implicit but this is better */
    } else if (is_def(line)) {
        
#line 528 "markmain.nw"
line = remove_def_marker(line);
while (*line && isspace(*line)) line++;
while (*line) {
  char tmp;
  char *s = line+1;
  while (*s && !isspace(*s)) s++;
  tmp = *s; *s = 0;
  print_index(out, Defn, line);
  *s = tmp;
  while (isspace(*s)) s++;
  line = s;
}
print_index(out, Newline, 0);
#line 137 "markmain.nw"
        if (state == Code)
            state = CodeIndex;
    } else {
        if (starts_doc(line) || state == CodeIndex) {
            
#line 542 "markmain.nw"
if (quoting) {
    errorat(filename, last_open_quote, Warning, "open quote `[[' never closed");
    quoting = 0;
}
#line 142 "markmain.nw"
            print_state(out, End, state, count);
            count++;
            state = Docs;       /* always reading docs after a stop */
            print_state(out, Begin, state, count);
            if (starts_doc(line))
                line = first_doc_line(line);
        }
        {   
#line 181 "markmain.nw"
char *p, *s, *t, c;
static char *buf;
static int buflen = 0;
#line 150 "markmain.nw"
            
#line 259 "markmain.nw"
#define LA1 '<'
#define LA2 '<'
#define RA1 '>'
#define RA2 '>'
#define LS1 '['
#define LS2 '['
#define RS1 ']'
#define RS2 ']'
#define ESC '@'
#line 272 "markmain.nw"
#define next(L) do {           c = *t++; goto L; } while (0)
#define copy(L) do { *s++ = c; c = *t++; goto L; } while (0)
#line 278 "markmain.nw"
#define lexassert(E) ((void)0)
#line 151 "markmain.nw"
            
#line 505 "markmain.nw"
if (buf == NULL)
  checkptr(buf = (char *) malloc(buflen = 128));
if (buflen < strlen(line) + 1 + 2) {
  while (buflen < strlen(line) + 1 + 2) 
    buflen *= 2;
  checkptr(buf = (char *) realloc(buf, buflen));
}
#line 186 "markmain.nw"
p = s = buf+2;
t = line;
c = *t++;
#line 152 "markmain.nw"
            
#line 206 "markmain.nw"
if (c == ESC && *t == ESC) { *s++ = ESC; c = *++t; ++t; }
#line 153 "markmain.nw"
            if (state == Code || quoting) 
                goto convert_code;
            else
                goto convert_docs;
            
#line 296 "markmain.nw"
convert_docs:
t: lexassert(state == Docs && !quoting);
   if (c == ESC) next(at);
   if (c == LA1) next(la);
   if (c == LS1) next(ls);
   
#line 242 "markmain.nw"
if (c == '\0') {
  
#line 513 "markmain.nw"
if (s > p) {
  *s = 0;
  print_pair(out, "text", p);
}
s = p = buf + 2;
#line 244 "markmain.nw"
  goto done_converting;
}
#line 302 "markmain.nw"
   copy(t);
#line 310 "markmain.nw"
at: if (c == LA1) next(atla);
    if (c == LS1) next(atls);
    if (c == RA1) next(atra);
    if (c == RS1) next(atrs);
    *s++ = ESC; goto t;
#line 319 "markmain.nw"
atls: if (c == LS2) { *s++ = LS1; *s++ = LS2; next(t); }
     *s++ = ESC; *s++ = LS1; goto t;
#line 325 "markmain.nw"
atla: if (c == LA2) { *s++ = LA1; *s++ = LA2; next(t); }
      *s++ = ESC; *s++ = LA1; goto t;
#line 331 "markmain.nw"
atrs: if (c == RS2) { *s++ = RS1; *s++ = RS2; next(t); }
     *s++ = ESC; *s++ = RS1; goto t;
#line 337 "markmain.nw"
atra: if (c == RA2) { *s++ = RA1; *s++ = RA2; next(t); }
      *s++ = ESC; *s++ = RA1; goto t;
#line 343 "markmain.nw"
la: if (c == LA2) {
      
#line 525 "markmain.nw"
errorat(filename, lineno, Error, "unescaped << in documentation chunk");
#line 345 "markmain.nw"
      *s++ = LA1; *s++ = LA2; next(t);
    }
    *s++ = LA1; goto t;
#line 355 "markmain.nw"
ls: lexassert(state == Docs);
    if (c == LS2) {
      
#line 513 "markmain.nw"
if (s > p) {
  *s = 0;
  print_pair(out, "text", p);
}
s = p = buf + 2;
#line 358 "markmain.nw"
      quoting = 1; last_open_quote = lineno; print_pair(out, "quote", 0); 
      next(c);
    }
    *s++ = LS1;
    goto t;
#line 372 "markmain.nw"
convert_code:
c: lexassert(state == Code || quoting);
   if (c == RS1 && quoting) next(crs);
   if (c == LA1) next(cla);
   if (c == ESC) next(cat);
   
#line 242 "markmain.nw"
if (c == '\0') {
  
#line 513 "markmain.nw"
if (s > p) {
  *s = 0;
  print_pair(out, "text", p);
}
s = p = buf + 2;
#line 244 "markmain.nw"
  goto done_converting;
}
#line 378 "markmain.nw"
   copy(c);
#line 386 "markmain.nw"
cat: if (c == LA1) next(catla);
     if (c == RA1) next(catra);
     *s++ = ESC; goto c;
#line 393 "markmain.nw"
catla: if (c == LA2) { *s++ = LA1; *s++ = LA2; next(c); }
       *s++ = ESC; *s++ = LA1; goto c;
#line 399 "markmain.nw"
catra: if (c == RA2) { *s++ = RA1; *s++ = RA2; next(c); }
       *s++ = ESC; *s++ = RA1; goto c;
#line 407 "markmain.nw"
crs: if (c == RS2) next(ce);
     *s++ = RS1; goto c;
#line 413 "markmain.nw"
ce: lexassert(quoting);
    if (c == RS2) copy(ce);
    quoting = 0; 
#line 513 "markmain.nw"
if (s > p) {
  *s = 0;
  print_pair(out, "text", p);
}
s = p = buf + 2;
#line 415 "markmain.nw"
                                                 print_pair(out, "endquote", 0);
    goto t;
#line 423 "markmain.nw"
cla: if (c == LA2) { 
#line 513 "markmain.nw"
if (s > p) {
  *s = 0;
  print_pair(out, "text", p);
}
s = p = buf + 2;
#line 423 "markmain.nw"
                                                     next(u); }
     *s++ = LA1; goto c;
#line 434 "markmain.nw"
u: lexassert(state == Code || quoting);
   if (c == LS1) copy(uls);
   if (c == RS1 && quoting) next(urs);
   if (c == RA1) next(ura);
   if (c == '\0') { /* premature end --- it's not a use after all */
     
#line 500 "markmain.nw"
lexassert(p == buf + 2);
p -= 2;
p[0] = LA1;
p[1] = LA2;
#line 440 "markmain.nw"
     goto c;
   }
   copy(u);
#line 447 "markmain.nw"
urs: lexassert(quoting);
     if (c == RS2) { /* premature end --- it's not a use after all */
       
#line 500 "markmain.nw"
lexassert(p == buf + 2);
p -= 2;
p[0] = LA1;
p[1] = LA2;
#line 450 "markmain.nw"
       next(ce); 
     }
     *s++ = RS1; goto u;
#line 457 "markmain.nw"
uls: if (c == LS2) copy(uc);
     goto u;
#line 464 "markmain.nw"
uc: lexassert(quoting || state == Code);
    if (c == RS1) copy(ucrs);
    if (c == '\0') {
      
#line 500 "markmain.nw"
lexassert(p == buf + 2);
p -= 2;
p[0] = LA1;
p[1] = LA2;
#line 468 "markmain.nw"
      goto c;
    }
    copy(uc);
#line 475 "markmain.nw"
ucrs: if (c == RS2) copy(uce);
      goto uc;
#line 481 "markmain.nw"
uce: if (c == RS2) copy(uce);
     goto u;
#line 487 "markmain.nw"
ura: if (c == RA2) { 
#line 520 "markmain.nw"
*s = 0;
print_pair(out, "use", p);
s = p = buf + 2;
#line 487 "markmain.nw"
                                                    next(c); }
     *s++ = RA1; goto u;
#line 158 "markmain.nw"
          done_converting: 
            (void)0;
        }
    }
}
#line 542 "markmain.nw"
if (quoting) {
    errorat(filename, last_open_quote, Warning, "open quote `[[' never closed");
    quoting = 0;
}
#line 174 "markmain.nw"
if (missing_newline) print_pair(out, "nl",0);
#line 165 "markmain.nw"
print_state(out, End, state, count);
#line 95 "markmain.nw"
}
#line 551 "markmain.nw"
main(int argc, char **argv) {
    FILE *fp;
    int i;

    progname = argv[0];	
    for (i = 1; i < argc && argv[i][0] == '-' && argv[i][1] != 0; i++)
        switch(argv[i][1]) {
            case 't': 
#line 582 "markmain.nw"
if (isdigit(argv[i][2]))
    tabsize = atoi(argv[i]+2);
else if (argv[i][2]==0)
    tabsize = 0;                /* no tabs */
else 
    errormsg(Error, "%s: ill-formed option %s", progname, argv[i]);
#line 558 "markmain.nw"
                                        break;
            default : errormsg(Error, "%s: unknown option -%c", progname, argv[i][1]);
                      break;
        }
    if (i < argc)
        for (; i < argc; i++) {
            if (!strcmp(argv[i], "-")) {
                markup(stdin,stdout,"");
            } else 
                if ((fp=fopen(argv[i],"r"))==NULL) {
                    errormsg(Error, "%s: couldn't open file %s", progname, argv[i]);
                    fprintf(stdout, "@fatal couldn't open file %s\n", argv[i]);
                } else {
                    markup(fp,stdout,argv[i]);
                    fclose(fp);
                }
        }
    else
        markup(stdin,stdout,"");
    nowebexit(NULL);
    return errorlevel;          /* slay warning */
}
